### PR TITLE
[BEAM-974] Add PubSub attributes support to PubsubIO.

### PR DESCRIPTION
**Pull request by reuvenlax**
Introduces a new PubsubMessage class, that allows separate management of payload and attributes from PubSub. Allows the user to register functions on PubsubIO to parse  and format these objects in order to interact with attributes.

### CODE REVIEW

**Comment by davorbonaci** [comment]
Thanks @reuvenlax.

Looks like a few test breakages here. Perhaps you can ping us when ready for review.

R: @davorbonaci @dhalperi

**Comment by reuvenlax** [comment]
Test breakages are fixed. Remaining failure is a known issue, unrelated to my PR.

**Comment by kennknowles** [comment]
There's actually an unused import causing it to fail fast on checkstyle.

**Comment by reuvenlax** [comment]
Everything now passes.

**Comment by reuvenlax** [comment]
@kennknowles do you have any comments?

**Comment by kennknowles** [comment]
Looks like this needs a rebase; I'm not sure what the conflicts might be but maybe I can take a look afterwards?

R: @kennknowles

**Comment by reuvenlax** [comment]
@dhalperi Should we try and get this in?

**Comment by dhalperi** [comment]
No

**Comment by dhalperi** [comment]
Put better, the 0.4.0-incubating release candidate is already cut and the vote is in progress. I'd encourage you to join the discussion on the dev mailing list. I'm sure you're not talking about anything else, because this would not be an appropriate venue for such a discussion.

As for the feature itself -- wonderful, but this code review for `gcp-io` is not likely to come to the top of my priority stack this week given the other Google-specific commitments we have. The Beam project is drawing lots of our resources in other parts of the stack -- I'm deprioritizing the Google-specific parts of Beam for now.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java [hunk 1]
```diff
@@ -1053,7 +1144,16 @@ private Bound(
        * @param <X> the type of the elements of the input {@link PCollection}
        */
       public <X> Bound<X> withCoder(Coder<X> coder) {
-        return new Bound<>(name, topic, timestampLabel, idLabel, coder);
+        return new Bound<>(name, topic, timestampLabel, idLabel, coder, null);
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java]: I don't understand why the parse function is dropped here. Seems like something fundamental is missing.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java [hunk 2]
```diff
@@ -1110,6 +1211,10 @@ public String getIdLabel() {
         return coder;
       }
 
+      public SimpleFunction<T, PubsubMessage> getFormatFn() {
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java]: javadoc all public functions

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSink.java [hunk 3]
```diff
@@ -419,6 +452,11 @@ public String getIdLabel() {
     return idLabel;
   }
 
+  @Nullable
+  public SimpleFunction<T, PubsubIO.PubsubMessage> getFormatFn() {
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSink.java]: javadoc public functions

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java [hunk 4]
```diff
@@ -152,6 +154,41 @@ private static void populateCommonDisplayData(DisplayData.Builder builder,
...
+     * Returns the main PubSub message.
+     */
+    public byte[] getMessage() {
+      return message;
+    }
+
+    /**
+     * Returns the given attribute value. If not such attribute exists, returns null.
+     */
+    public String getAttribute(String attribute) {
```
Comment (reuvenlax) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java]: done

Comment (reuvenlax) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java]: Done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java [hunk 5]
```diff
@@ -676,7 +730,20 @@ private Bound(String name, ValueProvider<PubsubSubscription> subscription,
...
+
+      /**
+       * Causes the source to return a PubsubMessage that includes Pubsub attributes.
+       * The user must supply a parsing function to transform the PubsubMessage into an output type.
+       * A Coder for the output type T must be registered or set on the output via
+       * {@link PCollection#setCoder(Coder)}.
+       */
+      public <T> Bound<T> withAttributes(SimpleFunction<PubsubMessage, T> parseFn) {
+        return new Bound<T>(
+            name, subscription, topic, timestampLabel, null, idLabel,
```
Comment (reuvenlax) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java]: Fixed

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java (hunk 1 above)
Comment (reuvenlax) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java]: Fixed

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java (hunk 2 above)
Comment (reuvenlax) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java]: done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSink.java (hunk 3 above)
Comment (reuvenlax) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubUnboundedSink.java]: done

**Comment by dhalperi** [comment]
You need to merge `master` in and fix up some conflicts. Looks like this PR is missing some improvements from Sam and to `TestPipeline`.

But looking pretty dang good.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java [hunk 6]
```diff
@@ -594,119 +512,168 @@ private Bound(String name, ValueProvider<PubsubSubscription> subscription,
...
+     *
+     * <p>See {@link PubsubIO.PubsubSubscription#fromPath(String)} for more details on the format
+     * of the {@code subscription} string.
+     *
+     * <p>Multiple readers reading from the same subscription will each receive
+     * some arbitrary portion of the data.  Most likely, separate readers should
+     * use their own subscriptions.
+     *
+     * <p>Does not modify this object.
+     */
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java]: seems like some Javadoc indent got offset by -2?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java [hunk 7]
```diff
@@ -594,119 +512,168 @@ private Bound(String name, ValueProvider<PubsubSubscription> subscription,
...
+     * Creates and returns a transform for reading from a Cloud Pub/Sub topic. Mutually exclusive
+     * with {@link #subscription(String)}.
+     *
+     * <p>See {@link PubsubIO.PubsubTopic#fromPath(String)} for more details on the format
+     * of the {@code topic} string.
+     *
+     * <p>The Beam runner will start reading data published on this topic from the time the pipeline
+     * is started. Any data published on the topic before the pipeline is started will not be read
+     * by the runner.
+     */
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java]: here, too

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java [hunk 8]
```diff
@@ -594,119 +512,168 @@ private Bound(String name, ValueProvider<PubsubSubscription> subscription,
...
+     /**
+      * Creates and returns a transform for reading from Cloud Pub/Sub where unique record
+      * identifiers are expected to be provided as Pub/Sub message attributes. The {@code idLabel}
+      * parameter specifies the attribute name. The value of the attribute can be any string
+      * that uniquely identifies this record.
+      *
+      * <p>Pub/Sub cannot guarantee that no duplicate data will be delivered on the Pub/Sub stream.
+      * If {@code idLabel} is not provided, Beam cannot guarantee that no duplicate data will
+      * be delivered, and deduplication of the stream will be strictly best effort.
+      */
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java]: and here.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java [hunk 9]
```diff
@@ -594,119 +512,168 @@ private Bound(String name, ValueProvider<PubsubSubscription> subscription,
...
-      public Bound<T> maxReadTime(Duration maxReadTime) {
-        return new Bound<>(
-            name, subscription, topic, timestampLabel, coder, idLabel, maxNumRecords, maxReadTime);
+     /**
+      * Creates and returns a transform for reading from Cloud Pub/Sub with a maximum number of
+      * records that will be read. The transform produces a <i>bounded</i> {@link PCollection}.
+      *
+      * <p>Either this option or {@link #maxReadTime(Duration)} must be set in order to create a
+      * bounded source.
+      */
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java]: and here, and a few more.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/PubsubClient.java [hunk 10]
```diff
@@ -310,6 +310,8 @@ public static TopicPath topicPathFromName(String projectId, String topicName) {
      */
     public final byte[] elementBytes;
 
+    public final Map<String, String> attributes;
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/PubsubClient.java]: javadoc

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/PubsubClient.java [hunk 11]
```diff
@@ -429,12 +438,13 @@ public boolean equals(Object o) {
              && requestTimeMsSinceEpoch == that.requestTimeMsSinceEpoch
              && ackId.equals(that.ackId)
              && recordId.equals(that.recordId)
-             && Arrays.equals(elementBytes, that.elementBytes);
+             && Arrays.equals(elementBytes, that.elementBytes)
+              && Objects.equal(attributes, that.attributes);
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/PubsubClient.java]: indent off.

**Comment by dhalperi** [comment]
LGTM. Assuming the next Jenkins run passes, happy to fixup minor spacing issues in merge.

**Comment by dhalperi** [comment]
Ugh, build failed with Javadoc broken links:

org.apache.maven.lifecycle.LifecycleExecutionException: Failed to
execute goal org.apache.maven.plugins:maven-javadoc-plugin:2.10.3:jar
(javadoc) on project beam-sdks-java-core: MavenReportException: Error
while generating Javadoc:
Exit code: 1 - /home/jenkins/jenkins-slave/workspace/beam_PreCommit_Java_MavenInstall/sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java:474:
warning - Tag @link: can't find maxNumRecords(int) in
org.apache.beam.sdk.io.Read
/home/jenkins/jenkins-slave/workspace/beam_PreCommit_Java_MavenInstall/sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java:474:
warning - Tag @link: can't find maxReadTime(Duration) in
org.apache.beam.sdk.io.Read
/home/jenkins/jenkins-slave/workspace/beam_PreCommit_Java_MavenInstall/sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java:1060:
error: unknown tag: linkValueProvider
     * Returns the {@linkValueProvider} for the topic being read from.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java (hunk 7 above)
Comment (reuvenlax) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java]: Done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java (hunk 6 above)
Comment (reuvenlax) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/PubsubIO.java]: Done

**Comment by dhalperi** [comment]
Here was my patch to fix it up. Don't you need a Coder in your example changes?

```diff
diff --git a/examples/java8/src/main/java/org/apache/beam/examples/complete/game/GameStats.java b/examples/java8/src/main/java/org/apache/beam/examples/complete/game/GameStats.java
index 74f1b304d..0ad09dbfa 100644
--- a/examples/java8/src/main/java/org/apache/beam/examples/complete/game/GameStats.java
+++ b/examples/java8/src/main/java/org/apache/beam/examples/complete/game/GameStats.java
@@ -24,6 +24,7 @@ import org.apache.beam.examples.common.ExampleUtils;
 import org.apache.beam.examples.complete.game.utils.WriteWindowedToBigQuery;
 import org.apache.beam.sdk.Pipeline;
 import org.apache.beam.sdk.PipelineResult;
+import org.apache.beam.sdk.coders.StringUtf8Coder;
 import org.apache.beam.sdk.io.PubsubIO;
 import org.apache.beam.sdk.options.Default;
 import org.apache.beam.sdk.options.Description;
@@ -252,7 +253,10 @@ public class GameStats extends LeaderBoard {
 
     // Read Events from Pub/Sub using custom timestamps
     PCollection<GameActionInfo> rawEvents = pipeline
-        .apply(PubsubIO.Read.timestampLabel(TIMESTAMP_ATTRIBUTE).topic(options.getTopic()))
+        .apply(PubsubIO.<String>read()
+            .timestampLabel(TIMESTAMP_ATTRIBUTE)
+            .topic(options.getTopic())
+            .withCoder(StringUtf8Coder.of()))
         .apply("ParseGameEvent", ParDo.of(new ParseEventFn()));
 
     // Extract username/score pairs from the event stream
diff --git a/examples/java8/src/main/java/org/apache/beam/examples/complete/game/LeaderBoard.java b/examples/java8/src/main/java/org/apache/beam/examples/complete/game/LeaderBoard.java
index 519bd5f53..253a80e6e 100644
--- a/examples/java8/src/main/java/org/apache/beam/examples/complete/game/LeaderBoard.java
+++ b/examples/java8/src/main/java/org/apache/beam/examples/complete/game/LeaderBoard.java
@@ -27,6 +27,7 @@ import org.apache.beam.examples.complete.game.utils.WriteToBigQuery;
 import org.apache.beam.examples.complete.game.utils.WriteWindowedToBigQuery;
 import org.apache.beam.sdk.Pipeline;
 import org.apache.beam.sdk.PipelineResult;
+import org.apache.beam.sdk.coders.StringUtf8Coder;
 import org.apache.beam.sdk.io.PubsubIO;
 import org.apache.beam.sdk.options.Default;
 import org.apache.beam.sdk.options.Description;
@@ -190,7 +191,10 @@ public class LeaderBoard extends HourlyTeamScore {
     // Read game events from Pub/Sub using custom timestamps, which are extracted from the pubsub
     // data elements, and parse the data.
     PCollection<GameActionInfo> gameEvents = pipeline
-        .apply(PubsubIO.Read.timestampLabel(TIMESTAMP_ATTRIBUTE).topic(options.getTopic()))
+        .apply(PubsubIO.<String>read()
+            .timestampLabel(TIMESTAMP_ATTRIBUTE)
+            .topic(options.getTopic())
+            .withCoder(StringUtf8Coder.of()))
         .apply("ParseGameEvent", ParDo.of(new ParseEventFn()));
 
     gameEvents.apply("CalculateTeamScores",
diff --git a/sdks/java/core/src/test/java/org/apache/beam/sdk/io/PubsubUnboundedSinkTest.java b/sdks/java/core/src/test/java/org/apache/beam/sdk/io/PubsubUnboundedSinkTest.java
index 30521e31c..24588d4ee 100644
--- a/sdks/java/core/src/test/java/org/apache/beam/sdk/io/PubsubUnboundedSinkTest.java
+++ b/sdks/java/core/src/test/java/org/apache/beam/sdk/io/PubsubUnboundedSinkTest.java
@@ -28,6 +28,7 @@ import java.util.List;
 import java.util.Map;
 
 import org.apache.beam.sdk.coders.StringUtf8Coder;
+import org.apache.beam.sdk.io.PubsubIO.PubsubMessage;
 import org.apache.beam.sdk.io.PubsubUnboundedSink.RecordIdMethod;
 import org.apache.beam.sdk.options.ValueProvider.StaticValueProvider;
 import org.apache.beam.sdk.testing.CoderProperties;
@@ -86,6 +87,13 @@ public class PubsubUnboundedSinkTest implements Serializable {
     CoderProperties.coderSerializable(PubsubUnboundedSink.CODER);
   }
 
+  private static class AttributeFn extends SimpleFunction<String, PubsubMessage> {
+    @Override
+    public PubsubMessage apply(String input) {
+      return new PubsubMessage(input.getBytes(), ATTRIBUTES);
+    }
+  }
+
   @Test
   @Category(NeedsRunner.class)
   public void sendOneMessage() throws IOException {
@@ -103,12 +111,7 @@ public class PubsubUnboundedSinkTest implements Serializable {
           new PubsubUnboundedSink<>(factory, StaticValueProvider.of(TOPIC), StringUtf8Coder.of(),
               TIMESTAMP_LABEL, ID_LABEL, NUM_SHARDS, batchSize, batchBytes,
               Duration.standardSeconds(2),
-              new SimpleFunction<String, PubsubIO.PubsubMessage>() {
-                @Override
-                public PubsubIO.PubsubMessage apply(String input) {
-                  return new PubsubIO.PubsubMessage(input.getBytes(), ATTRIBUTES);
-                }
-              },
+              new AttributeFn(),
               RecordIdMethod.DETERMINISTIC);
       p.apply(Create.of(ImmutableList.of(DATA)))
        .apply(ParDo.of(new Stamp()))
```

**Pull request closed without merging**